src/rgw/rgw_lc_tier.cc

```diff
@@ -0,0 +1,1242 @@
+        return set_cr_error(ret_err);
+      }
+
+      /* remove status obj */
+      ret_err = tier_ctx.store->delete_system_obj(status_obj.pool, status_obj.oid, nullptr, null_yield);
+
+      if (ret_err < 0) {
+        ldout(tier_ctx.cct, 0) << "ERROR: failed to abort multipart upload obj=" << tier_ctx.obj << " upload_id=" << status.upload_id << " part number " << status.cur_part << " (" << cpp_strerror(-ret_err) << ")" << dendl;
+        /* ignore error, best effort */
+      }
+      return set_cr_done();
+    }
+    return 0;
+  }
+};
+
+int RGWLCCloudCheckCR::operate() {
+  /* Check if object has already been transitioned */
+  rgw_lc_obj_properties obj_properties(tier_ctx.o.meta.mtime, tier_ctx.o.meta.etag,
+                                       tier_ctx.o.versioned_epoch, tier_ctx.acl_mappings,
+                                       tier_ctx.target_storage_class);
+
+  RGWBucketInfo b;
+  string target_obj_name;
+  int reterr = 0;
+
+  b.bucket.name = tier_ctx.target_bucket_name;
+  target_obj_name = tier_ctx.bucket_info.bucket.name + "/" +
+                    (*tier_ctx.obj)->get_name();
+    if (!tier_ctx.o.is_current()) {
+      target_obj_name += get_key_instance((*tier_ctx.obj)->get_key());
+    }
+
+  std::unique_ptr<rgw::sal::RGWBucket> dest_bucket;
+  std::unique_ptr<rgw::sal::RGWObject> dest_obj;
+
+
+  reterr = tier_ctx.store->get_bucket(nullptr, b, &dest_bucket);
+  if (reterr < 0) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_bucket - " << tier_ctx.target_bucket_name << " , reterr = " << reterr << dendl;
+    return reterr;
+  }
+  
+  dest_obj = dest_bucket->get_object(rgw_obj_key(target_obj_name));
+  if (!dest_obj) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_object path - " << target_obj_name << dendl;
+    return -1;
+  }
+
+  std::shared_ptr<RGWLCStreamGetCRF> get_crf;
+  get_crf.reset(new RGWLCStreamGetCRF((CephContext *)(tier_ctx.cct), get_env(), this,
+        (RGWHTTPManager*)(tier_ctx.http_manager),
+        obj_properties, tier_ctx.conn, static_cast<rgw::sal::RGWObject *>(dest_obj.get())));
+  int ret = 0;
+
+  reenter(this) {
+    /* Having yield here doesn't seem to wait for init2() to fetch the headers
+     * before calling is_already_tiered() below
+     */
+    ret = get_crf->init();
+    if (ret < 0) {
+      ldout(tier_ctx.cct, 0) << "ERROR: failed to fetch HEAD from cloud for obj=" << tier_ctx.obj << " , ret = " << ret << dendl;
+      return set_cr_error(ret);
+    }
+    if ((static_cast<RGWLCStreamGetCRF *>(get_crf.get()))->is_already_tiered()) {
+      *already_tiered = true;
+      ldout(tier_ctx.cct, 20) << "is_already_tiered true" << dendl;
+      return set_cr_done(); 
+    }
+
+    ldout(tier_ctx.cct, 20) << "is_already_tiered false..going with out_crf writing" << dendl;
+
+    return set_cr_done();
+  }
+  return 0;
+}
+
+map <pair<string, string>, utime_t> target_buckets;
```

Comment:
a global variable? does anything ever expire these entries over the lifetime of the process?

Reply:
yes this is to avoid sending bucket creation request for every object processed (unlike in cloud-sync case). The entries are considered expired post '2*lc_debug_interval' time interval. Hope that will suffice. 
However there is no background thread to delete such entries.